lib/rubocop/cop/performance/regexp_match.rb

```diff
@@ -0,0 +1,185 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Performance
+      # In Ruby 2.4, `String#match?` and `Regexp#match?` have been added.
+      # The methods are faster than `match`. Because the methods avoid
+      # creating a `MatchData` object or saving backref.
+      # So, when `MatchData` is not used, use `match?` instead of `match`.
+      #
+      # @example
+      #   # bad
+      #   def foo
+      #     if x =~ /re/
+      #       do_something
+      #     end
+      #   end
+      #
+      #   # bad
+      #   def foo
+      #     if x.match(/re/)
+      #       do_something
+      #     end
+      #   end
+      #
+      #   # good
+      #   def foo
+      #     if x.match?(/re/)
+      #       do_something
+      #     end
+      #   end
+      #
+      #   # good
+      #   def foo
+      #     if x =~ /re/
+      #       do_something(Regexp.last_match)
+      #     end
+      #   end
+      #
+      #   # good
+      #   def foo
+      #     if x.match(/re/)
+      #       do_something($~)
+      #     end
+      #   end
+      class RegexpMatch < Cop
+        MSG =
+          'Use `match?` instead of `%s` when `MatchData` is not used.'.freeze
+
+        def_node_matcher :match_method?, <<-PATTERN
+          {
+            (send _recv :match _)
+            (send _recv :match _ (:int ...))
+          }
+        PATTERN
+
+        def_node_matcher :match_operator?, <<-PATTERN
+          (send !nil :=~ !nil)
+        PATTERN
+
+        def_node_matcher :match_with_lvasgn?, <<-PATTERN
+          (match_with_lvasgn !nil !nil)
+        PATTERN
+
+        MATCH_NODE_PATTERN = <<-PATTERN.freeze
+          {
+            #match_method?
+            #match_operator?
+            #match_with_lvasgn?
+          }
+        PATTERN
+
+        def_node_matcher :match_node?, MATCH_NODE_PATTERN
+        def_node_search :search_match_nodes, MATCH_NODE_PATTERN
+
+        def_node_search :when_clauses, <<-PATTERN
+          (when ...)
+        PATTERN
+
+        def_node_search :last_matches, <<-PATTERN
+          {
+            (send (const nil :Regexp) :last_match)
+            (send (const nil :Regexp) :last_match _)
+            ({back_ref nth_ref} _)
+            (gvar #dollar_tilde)
+          }
+        PATTERN
+
+        def on_if(node)
+          return if target_ruby_version < 2.4
```

Comment:
Probably we simply need to add to the Cop class some method like `minimum_target_ruby_version` or something. I hate having to write such checks all over the place.